synapse/http/proxyagent.py

```diff
@@ -121,11 +122,11 @@ def __init__(
         self.https_proxy_creds, https_proxy = parse_username_password(https_proxy)
```

Comment:
Is there a reason why only credentials were parsed for https proxy? Only security reason or anything else?

Reply:
> this will break for `https://username:password@host`, I think? Probably best to call `parse_proxy` here

I would suggest a solution like this.
This change can make `parse_username_password` compatible to  `https://username:password@host`
```
diff --git a/synapse/http/proxyagent.py b/synapse/http/proxyagent.py
index c747b99..3c4fee6 100644
--- a/synapse/http/proxyagent.py
+++ b/synapse/http/proxyagent.py
@@ -293,7 +293,7 @@
 def parse_username_password(proxy: bytes) -> Tuple[Optional[ProxyCredentials], bytes]:
     """
     Parses the username and password from a proxy declaration e.g
-    username:password@hostname:port.
+    username:password@hostname:port or https://username:password@hostname:port
 
     Args:
         proxy: The proxy connection string.
@@ -304,9 +304,15 @@
         ProxyCredentials instance is replaced with None.
     """
     if proxy and b"@" in proxy:
+        scheme, host, port = parse_proxy(proxy)
         # We use rsplit here as the password could contain an @ character
-        credentials, proxy_without_credentials = proxy.rsplit(b"@", 1)
-        return ProxyCredentials(credentials), proxy_without_credentials
+        credentials, proxy_without_credentials = host.rsplit(b"@", 1)
+        return (
+            ProxyCredentials(credentials),
+            b"".join(
+                [scheme, b"://", proxy_without_credentials, b":", str(port).encode()]
+            ),
+        )
 
     return None, proxy
```

Reply:
> > Is there a reason why only credentials were parsed for https proxy? Only security reason or anything else?
> 
> I don't think there is much reason at all. Support for credentials was added in #9657: perhaps `http_proxy` was just fogotten?

In addition the issue for this #9000